src/DGtal/geometry/surfaces/DigitalSurfaceRegularization.ih

```diff
@@ -0,0 +1,422 @@
+{
+  return myDigitalSurface.isValid();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Implementation of inline functions                                        //
+///////////////////////////////////////////////////////////////////////////////
+template <typename T>
+inline
+void DGtal::DigitalSurfaceRegularization<T>::attachTrivialNormalVectors(const Parameters someParams)
+{
+  ASSERT_MSG(myInit, "The init() method must be called before setting the normals");
+  myNormals.clear();
+  myNormals = SHG3::getCTrivialNormalVectors( myDigitalSurface, mySurfels, someParams);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+template <typename T>
+inline
+void
+DGtal::DigitalSurfaceRegularization<T>::attachNormalVectors(const std::function<SHG3::RealVector(SH3::SCell&)> &normalFunc)
+{
+  ASSERT_MSG(myInit, "The init() method must be called before setting the normals");
+  myNormals.clear();
+  myNormals.resize( mySurfels.size());
+  for(auto i=0; i< mySurfels.size();++i)
+    myNormals[i] = normalFunc( mySurfels[i] );
+}
+///////////////////////////////////////////////////////////////////////////////
+//
+template <typename T>
+inline
+void DGtal::DigitalSurfaceRegularization<T>::cacheInit()
+{
+  //Collecting the surfels range
+  auto params  = SH3::defaultParameters() | SHG3::defaultParameters();
+  mySurfels.clear();
+  mySurfels = SH3::getSurfelRange( myDigitalSurface, params );
+  
+  myPointelIndex.clear();
+  //Collecting the input points
+  auto pointels        = SH3::getPointelRange(myPointelIndex,myDigitalSurface);
+  auto embedder        = SH3::getCellEmbedder(myK);  // /!\ no grid step here
+  myOriginalPositions.clear();
+  std::transform(pointels.begin(), pointels.end(), std::back_inserter(myOriginalPositions),
+                 [embedder](SH3::Cell &pointel ) {return embedder(pointel);  } );
+  
+  //Init the regularized positions.
+  myRegularizedPositions.resize(myOriginalPositions.size());
+  std::copy(myOriginalPositions.begin(), myOriginalPositions.end(), myRegularizedPositions.begin());
+  
+  //Allocating Gradient vector
+  myGradient.resize(myOriginalPositions.size());
+  myGradient.clear();
+  myGradientAlign.resize(myOriginalPositions.size());
+  myGradient.clear();
+  
+  /////
+  ///Cacheing some topological information
+  auto polySurf        = SH3::makeDualPolygonalSurface(mySurfelIndex,myDigitalSurface);
+  myFaces = polySurf->allFaces() ;  //All faces of the dual digital surface
+  auto dsurf_faces = myDigitalSurface->allClosedFaces(); // Faces of digital surface (umbrellas)
+  // dsurf_pointels list the pointels in the same order as the faces of polySurf.
+  std::vector<SH3::Cell> dsurf_pointels( dsurf_faces.size() );
+  std::transform( dsurf_faces.cbegin(), dsurf_faces.cend(),
+                 dsurf_pointels.begin(),
+                 [&] ( const SH3::DigitalSurface::Face f ) { return myK.unsigns(myDigitalSurface->pivot( f )); } );
```

Comment:
I am not sure that this way of building pointers will give the same order as getPointelRange.

Reply:
Have you checked that the ordering of pointels is the same as getPointelRange ?

Reply:
Not yet...

Reply:
Pointels ordering looks fine even with your PR.